subsys/net/lib/dns/llmnr_responder.c

```diff
@@ -0,0 +1,610 @@
+		goto quit;
+	}
+
+	dns_msg.msg = dns_data->data;
+	dns_msg.msg_size = data_len;
+
+	ret = llmnr_unpack_query_header(&dns_msg, &dns_id);
+	if (ret < 0) {
+		ret = -EINVAL;
+		goto quit;
+	}
+
+	queries = ret;
+
+	NET_DBG("Received %d %s from %s (id 0x%04x)", queries,
+		queries > 1 ? "queries" : "query",
+		net_pkt_family(pkt) == AF_INET ?
+		net_sprint_ipv4_addr(&NET_IPV4_HDR(pkt)->src) :
+		net_sprint_ipv6_addr(&NET_IPV6_HDR(pkt)->src), dns_id);
+
+	do {
+		enum dns_rr_type qtype;
+		enum dns_class qclass;
+
+		memset(result->data, 0, net_buf_tailroom(result));
+		result->len = 0;
+
+		ret = dns_unpack_query(&dns_msg, result, &qtype, &qclass);
+		if (ret < 0) {
+			goto quit;
+		}
+
+		NET_DBG("[%d] query %s/%s label %s (%d bytes)", queries,
+			qtype == DNS_RR_TYPE_A ? "A" : "AAAA", "IN",
+			result->data, ret);
+
+		/* If the query matches to our hostname, then send reply */
+		if (!strncasecmp(hostname, result->data + 1, hostname_len) &&
+		    (result->len - 1) >= hostname_len) {
+			NET_DBG("LLMNR query to our hostname %s", hostname);
+			send_response(ctx, pkt, result, qtype, dns_id);
+		}
+	} while (--queries);
+
+	ret = 0;
+
+quit:
+	if (result) {
+		net_pkt_frag_unref(result);
+	}
+
+	return ret;
+}
+
+static void recv_cb(struct net_context *net_ctx,
+		    struct net_pkt *pkt,
+		    int status,
+		    void *user_data)
+{
+	struct net_context *ctx = user_data;
+	struct net_buf *dns_data = NULL;
+	struct dns_addrinfo info = { 0 };
+	int ret;
+
+	ARG_UNUSED(net_ctx);
+	NET_ASSERT(ctx == net_ctx);
+
+	if (!pkt) {
+		return;
+	}
+
+	if (status) {
+		goto quit;
+	}
+
+	dns_data = net_buf_alloc(&llmnr_dns_msg_pool, BUF_ALLOC_TIMEOUT);
+	if (!dns_data) {
+		goto quit;
+	}
+
+	ret = dns_read(ctx, pkt, dns_data, &info);
+	if (ret < 0 && ret != -EINVAL) {
+		NET_DBG("LLMNR read failed (%d)", ret);
+	}
+
+	net_buf_unref(dns_data);
+
+quit:
+	net_pkt_unref(pkt);
+}
+
+#if defined(CONFIG_NET_IPV6)
+static void iface_ipv6_cb(struct net_if *iface, void *user_data)
+{
+	struct in6_addr *addr = user_data;
+	int ret;
+
+	ret = net_ipv6_mld_join(iface, addr);
+	if (ret < 0) {
+		NET_DBG("Cannot join %s IPv6 multicast group (%d)",
+			net_sprint_ipv6_addr(addr), ret);
+	}
+}
+
+static void setup_ipv6_addr(struct sockaddr_in6 *local_addr)
+{
+	create_ipv6_addr(local_addr);
+
+	net_if_foreach(iface_ipv6_cb, &local_addr->sin6_addr);
+}
+#endif /* CONFIG_NET_IPV6 */
+
+#if defined(CONFIG_NET_IPV4)
+static void iface_ipv4_cb(struct net_if *iface, void *user_data)
+{
+	struct in_addr *addr = user_data;
+	struct net_if_mcast_addr *ifaddr;
+
+	ifaddr = net_if_ipv4_maddr_add(iface, addr);
+	if (!ifaddr) {
+		NET_DBG("Cannot add IPv4 multicast address to iface %p",
+			iface);
+	}
+}
+
+static void setup_ipv4_addr(struct sockaddr_in *local_addr)
+{
+	create_ipv4_addr(local_addr);
+
+	net_if_foreach(iface_ipv4_cb, &local_addr->sin_addr);
+}
+#endif /* CONFIG_NET_IPV4 */
+
+static int init_listener(void)
+{
+	int ret, ok = 0;
+
+#if defined(CONFIG_NET_IPV6)
+	do {
```

Comment:
why the do { } while() ? {} would be sufficient. Moreover you use a goto (and not a break) inside. Either way: do/while + break or {} + goto.